packages/cli/src/ui/hooks/vim.ts

```diff
@@ -0,0 +1,1420 @@
+    },
+    [buffer.lines],
+  );
+
+  /** Handles change operations for directional movement (ch, cj, ck, cl) */
+  const handleChangeMovement = useCallback(
+    (movementType: 'h' | 'j' | 'k' | 'l', count: number) => {
+      const currentRow = buffer.cursor[0];
+      const currentCol = buffer.cursor[1];
+
+      switch (movementType) {
+        case 'h': {
+          // Change N characters to the left
+          for (let i = 0; i < count; i++) {
+            if (currentCol > 0) {
+              buffer.move('left');
+              buffer.del();
+            }
+          }
+          break;
+        }
+
+        case 'j': {
+          // Change from current line down N lines
+          const totalLines = buffer.lines.length;
+          const linesToChange = Math.min(count + 1, totalLines - currentRow);
+
+          if (totalLines === 1) {
+            const currentLine = buffer.lines[0] || '';
+            buffer.replaceRangeByOffset(0, currentLine.length, '');
+          } else {
+            const { startOffset, endOffset } = getLineRangeOffsets(
+              currentRow,
+              linesToChange,
+            );
+            buffer.replaceRangeByOffset(startOffset, endOffset, '');
+          }
+          break;
+        }
+
+        case 'k': {
+          // Change from current line up N lines
+          const linesToChange = Math.min(count + 1, currentRow + 1);
+          const startRow = currentRow - count;
+
+          if (buffer.lines.length === 1) {
+            const currentLine = buffer.lines[0] || '';
+            buffer.replaceRangeByOffset(0, currentLine.length, '');
+          } else {
+            const { startOffset, endOffset } = getLineRangeOffsets(
+              Math.max(0, startRow),
+              linesToChange,
+            );
+            buffer.replaceRangeByOffset(startOffset, endOffset, '');
+            buffer.moveToOffset(startOffset);
+          }
+          break;
+        }
+
+        case 'l': {
+          // Change N characters to the right
+          for (let i = 0; i < count; i++) {
+            buffer.del();
+          }
+          break;
+        }
+
+        default:
+          // This should never happen due to type constraints
+          break;
+      }
+
+      dispatch({ type: 'SET_MODE', mode: 'INSERT' });
+      const cmdTypeMap = {
+        h: CMD_TYPES.CHANGE_MOVEMENT.LEFT,
+        j: CMD_TYPES.CHANGE_MOVEMENT.DOWN,
+        k: CMD_TYPES.CHANGE_MOVEMENT.UP,
+        l: CMD_TYPES.CHANGE_MOVEMENT.RIGHT,
+      } as const;
+
+      dispatch({
+        type: 'SET_LAST_COMMAND',
+        command: { type: cmdTypeMap[movementType], count },
+      });
+      dispatch({ type: 'SET_PENDING_C', pending: false });
+    },
+    [buffer, getLineRangeOffsets, dispatch],
+  );
+
+  /** Handles end-of-line operations (D, C) */
+  const handleEndOfLineOperation = useCallback(
+    (shouldEnterInsertMode: boolean) => {
+      const currentRow = buffer.cursor[0];
```

Comment:
the conditional logic in this method needs to go into textBuffer as it will otherwise have the stale issues discussed above